crates/environments/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+        .context("Internal error constructing WAC document for target checking")?;
+
+    // TODO: if we end up needing the registry, we need to do this dance
+    // for things we are providing separately, or the registry will try to
+    // hoover them up and will fail.
+    // let mut refpkgs = wac_resolver::packages(&doc)?;
+    // refpkgs.shift_remove(&env_pkg_key);
+    // refpkgs.shift_remove(&component_key);
+
+    // TODO: determine if this is needed in circumstances other than the simple test
+    // let reg_resolver = wac_resolver::RegistryPackageResolver::new(Some("wa.dev"), None).await?;
+    // let mut packages = reg_resolver
+    //     .resolve(&refpkgs)
+    //     .await
+    //     .context("reg_resolver.resolve failed")?;
+
+    let mut packages: indexmap::IndexMap<wac_types::BorrowedPackageKey, Vec<u8>> =
+        Default::default();
+
+    packages.insert(env_pkg_key, env.package_bytes().to_vec());
+    packages.insert(component_key, component.wasm_bytes().to_vec());
+
+    match doc.resolve(packages) {
+        Ok(_) => Ok(()),
+        Err(wac_parser::resolution::Error::TargetMismatch { kind, name, world, .. }) => {
+            // This one doesn't seem to get hit at the moment - we get MissingTargetExport or ImportNotInTarget instead
+            Err(anyhow!("Component {} ({}) can't run in environment {env_name} because world {world} expects an {} named {name}", component.id(), component.source_description(), kind.to_string().to_lowercase()))
+        }
+        Err(wac_parser::resolution::Error::MissingTargetExport { name, world, .. }) => {
+            Err(anyhow!("Component {} ({}) can't run in environment {env_name} because world {world} requires an export named {name}, which the component does not provide", component.id(), component.source_description()))
+        }
+        Err(wac_parser::resolution::Error::PackageMissingExport { export, .. }) => {
+            // TODO: The export here seems wrong - it seems to contain the world name rather than the interface name
+            Err(anyhow!("Component {} ({}) can't run in environment {env_name} because world {target_world} requires an export named {export}, which the component does not provide", component.id(), component.source_description()))
+        }
+        Err(wac_parser::resolution::Error::ImportNotInTarget { name, world, .. }) => {
+            Err(anyhow!("Component {} ({}) can't run in environment {env_name} because world {world} does not provide an import named {name}, which the component requires", component.id(), component.source_description()))
+        }
+        Err(e) => {
+            Err(anyhow!(e))
+        },
+    }
+}
+
+/// Equivalent to futures::future::join_all, but specialised for iterators of
+/// fallible futures. It returns a Result<Vec<...>> instead of a Vec<Result<...>> -
+/// this just moves the transposition boilerplate out of the main flow.
+async fn join_all_result<T, I>(iter: I) -> anyhow::Result<Vec<T>>
```

Comment:
Since we only care about the first error wouldn't https://docs.rs/futures/0.3.30/futures/future/fn.try_join_all.html be a better fit here? As its currently implemented, we wait for all futures to finish even though we only ever look at the first error.

Reply:
Just testing you, Wilson.

![image](https://github.com/user-attachments/assets/12731736-a44a-4baf-ad72-10cb5cf92ebb)

(Thanks.  I had completely forgotten about this and yes, it is exactly what I want.  Custom function removed.)